### PR TITLE
BUG: Title sometimes is bytes and not str.

### LINKED ISSUE #2929: `DocumentInformation.title` sometimes return `bytes` instead of `str`
I am reading a PDF file from:
https://www.ms-ad-hd.com/en/ir/ir_event/event/presentation/main/01111119/teaserItems1/00/linkList/00/link/20220810Tranverse%20QA%20Summary.pdf
The "title" for this doc returns `bytes` instead of `str`, although the method should always return `str`.

## Environment

Python 3.11

```bash
$ python -m platform
macOS-14.4.1-arm64-arm-64bit

$ python -c "import pypdf;print(pypdf._debug_versions)"
pypdf==4.0.1, crypt_provider=('cryptography', '41.0.7'), PIL=9.5.0
```

## Code + PDF

```python
import io
import pypdf
import requests

response = requests.get('https://www.ms-ad-hd.com/en/ir/ir_event/event/presentation/main/01111119/teaserItems1/00/linkList/00/link/20220810Tranverse%20QA%20Summary.pdf')
pdf_reader = pypdf.PdfReader(io.BytesIO(response.content))
print(pdf_reader.metadata.title)
```

### PR DESCRIPTION
**Pull request by reformy**
closes #2929

### CODE REVIEW

**Comment by stefan6419846** [comment]
Thanks for the PR. I assume that you own all the necessary copyrights for the test file?

For me this mostly looks okay, apart from the review comment left above about the position after the comment. Nevertheless, I am going to wait for @pubpub-zz here to have another look.

**Comment by pubpub-zz** [comment]
Personnaly I would have prefer to add the __str__ to the BytesStringObject() with trying diffrerent encoding(see §3.8.1) .

we should use CHARSET (https://github.com/py-pdf/pypdf/blob/98aa9742e757ec428e1953ba7f47c6d7c44b331a/pypdf/generic/_base.py#L841) adding as first format UTF-16 (based on the spec)

It would then just require to add a call to str in the diffrerent properties of metadata in https://github.com/py-pdf/pypdf/blob/main/pypdf/_doc_common.py

**Comment by reformy** [comment]
Thank you for your comments.
I've moved the utf-8 to be after the pdfdocencoding.
@pubpub-zz I did not understand your comment. If the whole str-decoding is not in the right place, that a different issue. I've just added another option to decode it.

**Comment by stefan6419846** [comment]
> If the whole str-decoding is not in the right place, that a different issue.

No, this is related to this issue. The suggestion is to revert the current changes and provide `ByteStringObject.__str__()` accordingly, which would then test the different charsets. At the same time, `DocumentInformation._get_text()` should be rewritten to check for `StringObject` and `ByteStringObject` and - instead of returning the pure value - return `str(retval)`.

**Comment by reformy** [comment]
I've changed the code as suggested, adding `__str__` and calling it from `_get_text`. This works well.

However, adding `utf-16` to the `CHARSETS` list breaks other things. I don't understand why do we need this and how is this related to my fix.

File: pypdf/generic/_utils.py [hunk 1]
```diff
@@ -189,6 +189,15 @@ def create_string_object(
                 retval.autodetect_pdfdocencoding = True
                 return retval
             except UnicodeDecodeError:
+                # Try utf-8.
```
Comment (stefan6419846) [inline pypdf/generic/_utils.py:192]: Isn't this unnecessary now?

**Comment by stefan6419846** [comment]
> However, adding utf-16 to the CHARSETS list breaks other things. I don't understand why do we need this and how is this related to my fix.

This is another encoding to consider. You should only consider this in `__str__`, not add it to the list permanently.

File: pypdf/generic/_base.py [hunk 2]
```diff
@@ -619,6 +619,15 @@ def write_to_stream(
         stream.write(binascii.hexlify(self))
         stream.write(b">")
 
+    def __str__(self) -> str or None:
+        for enc in NameObject.CHARSETS:
+            try:
+                ret = self.decode(enc)
+                return NameObject(ret)
```
Comment (stefan6419846) [inline pypdf/generic/_base.py:626]: Why do we need to wrap this into a `NameObject`?

File: pypdf/generic/_base.py [hunk 3]
```diff
@@ -619,6 +619,15 @@ def write_to_stream(
...
+            except Exception:
```
Comment (stefan6419846) [inline pypdf/generic/_base.py:627]: Which actual exceptions can occur here? `Exception` is rather broad and I would have expected to see only `UnicodeDecodeError` here.

File: pypdf/generic/_base.py [hunk 4]
```diff
@@ -619,6 +619,15 @@ def write_to_stream(
...
         stream.write(b">")
 
+    def __str__(self) -> str:
+        charset_to_try = ["utf-16"] + list(NameObject.CHARSETS)
+        for enc in charset_to_try:
+            try:
+                return self.decode(enc)
+            except UnicodeDecodeError:
+                pass
+        return ""
```
Comment (stefan6419846) [inline pypdf/generic/_base.py:629]: Shouldn't this rather raise an exception? For now, I consider it misleading to return an empty string for non-empty inputs.

Comment (reformy) [reply pypdf/generic/_base.py:629]: Again, the calling code (`DocumentInformation.title`) assumes that `_get_text` will return `''` if failed to decode, and I didn't want to change any unrelated behavior other than the bug I am fixing. But it does make sense, so I will change and hopefully no test will fail.

**Pull request merged**